src/lib/pipes/normalise-schedule-pipe.js

```diff
@@ -0,0 +1,263 @@
+                }
+
+                // Get schedule properties that will carry over to the generated events
+                let eventDataBase = {...parent};
+                delete eventDataBase["@id"];
+                eventDataBase["@type"] = schedule.scheduledEventType ? schedule.scheduledEventType : childKind;
+                eventDataBase.duration = schedule.duration;
+                eventDataBase.startDate = schedule.startDate;
+                eventDataBase.endDate = schedule.endDate;
+                eventDataBase.startTime = schedule.startTime;
+                eventDataBase.endTime = schedule.endTime;
+
+                // Get schedule properties used to calculate the series
+                try{
+                    let {freq, interval} = this.freq(schedule);
+                    let byDay = this.byWeekDay(schedule);
+                    let byMonth = this.byMonth(schedule);
+                    let byMonthDay = this.byMonthDay(schedule);
+                    let dtStart = this.dtStart(schedule);
+                    let until = this.until(schedule);
+                    let count = this.count(schedule);
+
+                    let rruleOptions = {freq: freq, interval: interval}; // this is the only required one
+                    if(typeof(byDay) !== 'undefined'){
+                        rruleOptions.byweekday = byDay;
+                    }
+                    if(typeof(byMonth) !== 'undefined'){
+                        rruleOptions.bymonth = byMonth;
+                    }
+                    if(typeof(byMonthDay) !== 'undefined'){
+                        rruleOptions.bymonthday = byMonthDay;
+                    }
+                    if(typeof(dtStart) !== 'undefined'){
+                        rruleOptions.dtstart = dtStart;
+                    }
+                    if(typeof(until) !== 'undefined'){
+                        rruleOptions.until = until;
+                    }
+                    if(typeof(count) !== 'undefined'){
+                        rruleOptions.count = count;
+                    }
+
+                    const rruleSet = new RRule.RRuleSet();
+                    const rule = new RRule.RRule(rruleOptions);
+                    rruleSet.rrule(rule);
+
+                    // Add dates to exclude if present
+                    // TODO: this only excludes exact matches, including time.
+                    //       It should also let input be yyyy-mm-dd and exclude any events that
+                    //       start in that 24 hour period.
+                    if(typeof schedule.exceptDate !== 'undefined'){
+                        const exceptDates = Utils.ensureArray(schedule.exceptDate);
+                        for(let exDate of exceptDates){
+                            rruleSet.exdate(new Date(exDate));
+                        }
+                    }
+
+                    // Only generate events two weeks from NOW (or until schedule ends, whichever is sooner)
```

Comment:
That's interesting.

If an thing is published for 2 months starting 1st Aug on the 1st Aug, we get the data 2nd Aug and make events till 15th Aug, and then it's never updated again ..... won't the system never generate normalised events for the rest of Aug and Sept?

(If you think this should be issued and looked at later happy for this not to be a blocker)

Reply:
Maybe the right thing to do here is generate all of them, then filter it down to two weeks from now in the RPDE output at query time?

Reply:
Is there any reason not to generate and RPDE them all straight up? Are we just worried about a bad bit of data declaring an event every hour for a year and filling the database or something?

Reply:
I have spun this out into https://github.com/openactive/conformance-services/issues/93